### PR TITLE
Implement coercers

### PR DESCRIPTION
**Pull request by pignatenko**
## TL;DR

The point of this PR is to create syntactic sugar for handling Typescript's null strictness in a safe way to prevent pedantic mistakes for mixing up `null` and `undefined` when it doesn't truly matter.

## Empty type

We create a new type called `Empty` exposed as a `t.empty` converter which allows you to pass `undefined` or `null` values.

## Coercers

Building on this we create a concept called a `coercer` which is exposed as `t.coerce` this allows you to input a set of converters and a value, if any converter successfully passes it returns the given value.

We bake in 2 default coercers called `emptyAsNull` and `emptyAsUndefined`. These coercers allow `empty` values but return `null` and `undefined` respectfully.

## CONTROVERSIAL SECTION -- optional

The idea here is to then pass on the `empty` concept into `t.optional`.

This PR modifies `t.optional` to coerce `null` into `undefined` and creates a `t.strictOptional` which retains the original behavior.

This is arguably the most semantic way to handle this and preserves the original intent that this library is an `undefined` based library following the MS standard but can seamlessly interoperate with libraries / datasets that follow a null standard or a mixed standard.

This is probably the most controversial change here, there are obviously other solutions to organizing this. Let me know what you guys think.

### CODE REVIEW

File: src/basic-types.ts [hunk 1]
```diff
@@ -96,11 +96,15 @@ const NullConverter = literal<null>(null);
  */
 const UndefinedConverter = literal<undefined>(undefined);
 
+export type Empty = null | undefined;
+const EmptyConverter: Converter<Empty> = NullConverter.or(UndefinedConverter);
```
Comment (RoboCafaz) [inline src/basic-types.ts:100]: The name `Empty` has me thinking about `[]` and `{}`. Is there a better name for `null | undefined`?

Comment (pignatenko) [reply src/basic-types.ts:100]: `nothing`?

Comment (pignatenko) [reply src/basic-types.ts:100]: Also `none`

Comment (RoboCafaz) [reply src/basic-types.ts:100]: Any of those would work for me. Maybe put it out for a poll on #dev?

File: src/coercers.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    if (converters.some((c) => testConverter(c, input, path, entity))) {
+      return value;
+    }
+
+    throw new ConverterError(input, name, path);
+  }, name);
+};
+
+export const emptyAsNull: Converter<Empty> = coerce([empty], null);
+export const emptyAsUndefined: Converter<Empty> = coerce([empty], undefined);
```
Comment (RoboCafaz) [inline src/coercers.ts:41]: TSDocs pls

File: src/basic-types.ts (hunk 1 above)
Comment (qcharlieshi) [reply src/basic-types.ts:100]: We use `none` in the `optional` lib in core-lib

File: src/basic-types.ts [hunk 3]
```diff
@@ -96,11 +96,15 @@ const NullConverter = literal<null>(null);
  */
 const UndefinedConverter = literal<undefined>(undefined);
 
+export type Empty = null | undefined;
```
Comment (RoboCafaz) [inline src/basic-types.ts:99]: TSDocs pls

**Comment by RoboCafaz** [comment]
I'm okay with the baseline change to optional. Just make sure this is a major version release and it gets communicated in release notes.

**Review by RoboCafaz — APPROVED** [review]
Just naming/doc nits. Functionality looks good.

File: src/basic-types.ts (hunk 1 above)
Comment (pignatenko) [reply src/basic-types.ts:100]: @qcharlieshi That sounds like a good reason to go with `None`/`none` then.

**Comment by pignatenko** [comment]
> Just naming/doc nits. Functionality looks good.

TSDocs/naming fixed. Will wait for consensus regarding `optional`.

File: src/decorators.ts [hunk 4]
```diff
@@ -2,16 +2,33 @@ import { Converter, ConverterFunction, createConverter, getConverterName } from
...
+export function strictOptional<Result, Input = unknown>(
+  converter: ConverterFunction<Result, Input>
+): Converter<Result | undefined, Input | undefined> {
+  return createConverter((input, path, entity) => {
+    if (input === undefined) {
+      return undefined;
+    }
+    return converter(input, path, entity);
+  }, `strictOptional ${getConverterName(converter)}`);
+}
```
Comment (chris-pardy) [inline src/decorators.ts:34]: From a versioning perspective I'd prefer we went the other way and introduced a coerced optional rather than a changing the behavior of optional. In particular there are many places where we use `null` to mean "I don't know" and undefined to mean don't update this field. Something like: `{ downPayment?: null | number }` which would break in this case as the nulls would become undefindes.
Also in general I think this library should strive for a set of strict core types and then compose them into less strict types, the core boolean, string, and number converters don't do this and I think it was a mistake.

As far as making breaking changes go I'd like to avoid that if possible and just document them in issues instead. This whole library is due for a major facelift in V1. In particular there are just fundamental mistakes that need to be revisited like relying on thrown errors to indicate failures, and treating missing and undefined values as equivalent.

**Review by chris-pardy — CHANGES_REQUESTED** [review]

File: src/decorators.ts (hunk 4 above)
Comment (pignatenko) [reply src/decorators.ts:34]: So reverting `t.strictOptional` to `t.optional` and creating a `t.coercedOptional` would rectify your objections? I am also trying to avoid paperwork/process that comes with major version changes etc.

Comment (RoboCafaz) [reply src/decorators.ts:34]: I'm also okay with this.

Comment (chris-pardy) [reply src/decorators.ts:34]: Short answer: yes.
Long answer:
I agree with the idea of having this version of optional be the "normal version you use" but I think we should avoid bumping this library to V1 for this.

Comment (pignatenko) [reply src/decorators.ts:34]: Alright I'll change this up and we can put the breaking functionality as part of the issues for V1.

Comment (kstrickland0612) [reply src/decorators.ts:34]: I'm also in favor of maintaining `t.optional`'s current behavior and handling the coercion in a new converter.

File: src/coercers.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  const predicate = converters.map((v) => getConverterName(v)).join(' | ');
+  const name = `${predicate} -> ${displayValue(value)}`;
+
+  return createConverter<V>((input, path, entity): V => {
+    if (converters.some((c) => testConverter(c, input, path, entity))) {
+      return value;
+    }
+
+    throw new ConverterError(input, name, path);
+  }, name);
```
Comment (chris-pardy) [inline src/coercers.ts:37]: from an error readability perspective consider defining this as:
```ts
const predicate = converters.reducer((c1, c2) => c1.or(c2));
return predicate.pipe(() => value, displayValue(value));
```

This way the error thrown will be the one from the union, which tries it best to select a specific error.

File: src/coercers.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Given a set of converters if one succeeds at converting the input return the
+ * given value.
+ *
+ * @param converters - Array of converters to pass in.
+ * @param value - Value to return
+ *
+ */
+export const coerce = <V>(converters: Array<Converter<any>>, value: V): Converter<V> => {
```
Comment (chris-pardy) [inline src/coercers.ts:27]: I would type this as `converters: [ConverterFunction<any>, ...ConverterFunction<any>[]]` to enforce a minimum of 1 converter.

File: src/coercers.ts (hunk 5 above)
Comment (chris-pardy) [reply src/coercers.ts:37]: For something like `t.coerce([t.string, t.null], 'something')({ notString: true })` then the error will be `At $ expected string | null but was { notString: true }`. However for something like: `t.coerce([t.strict({ isString: t.boolean }), t.null], 'something')({ notString: true })` the error if you do the or joining of the input converters will be: `At $.isString expected boolean but was undefined`

File: src/decorators.ts (hunk 4 above)
Comment (chris-pardy) [reply src/decorators.ts:34]: @pignatenko feel free to comment on #25 with any ideas.

File: src/coercers.ts (hunk 6 above)
Comment (chris-pardy) [reply src/coercers.ts:27]: Yes, you would need an unwrap code but again I'm ok with that.

Comment (chris-pardy) [reply src/coercers.ts:27]: One thing to keep in mind is that `t.coerce(t.string, 3)` is behaviorally equivalent to `t.string.pipe(() => 3)`

Comment (pignatenko) [reply src/coercers.ts:27]: I ended up doing type enforcement as the shortest path forward. Array wrapping would be a useful addon, but I am unsure of the perf due to how we use errors, so to keep things speedy for now and to resolve how V1 looks explicit arrays it is.

File: src/coercers.ts (hunk 5 above)
Comment (pignatenko) [reply src/coercers.ts:37]: I end up doing this, the perf hit is similar in the general case, in the exceptional case the custom logic performs better but it's not worth it losing error resolution and adding maintinance.

File: src/decorators.ts (hunk 4 above)
Comment (pignatenko) [reply src/decorators.ts]: @kstrickland0612 @chris-pardy @RoboCafaz `optional` has been reverted and this functionality has become a "choose your own adventure" for now that maps to the coercer concepts.

`t.noneable` will pass `null` and `undefined` thru.
`t.noneableAsNull` will coerce to `null`.
`t.noneableAsUndefined` will coerce to `undefined`.

File: src/coercers.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * @param converters - Array of converters to pass in.
+ * @param value - Value to return
+ *
+ */
+export const coerce = <V>(
+  converters: [Converter<any>, ...Converter<any>[]],
+  value: V
+): Converter<V> => {
+  const predicate = converters.reduce((c1, c2) => c1.or(c2));
+  return predicate.pipe(() => value, `${predicate.name} -> ${displayValue(value)}`);
```
Comment (chris-pardy) [inline src/coercers.ts:18]: use `predicate.displayName` for the correct value.

File: src/coercers.ts (hunk 6 above)
Comment (pignatenko) [reply src/coercers.ts]: Had to bump rollup dts plugin because it was failing to parse TS4.2

**Review by chris-pardy — APPROVED** [review]
@pignatenko approving, but I'd love your take on how this plays into the whole conversation about:
`t.optional(t.string.default(() => 'empty'))` for instance if I did: `t.noneableAsNull(t.string.default(() => 'empty'))(undefined)` what would I get?

File: src/decorators.ts [hunk 8]
```diff
@@ -15,3 +16,56 @@ export function optional<Result, Input = unknown>(
...
+
+/**
+ * Given a converter function make an noneable converter function that also coerces
+ * `none` to `null`
+ * @param converter - the inner converter function
+ */
+export function noneableAsNull<Result, Input = unknown>(
+  converter: ConverterFunction<Result, Input>
+): Converter<Result | null, Input | undefined> {
```
Comment (chris-pardy) [inline src/decorators.ts:48]: I think these types are: `<Result | null, Input | null | undefined>`

File: src/decorators.ts [hunk 9]
```diff
@@ -15,3 +16,56 @@ export function optional<Result, Input = unknown>(
...
+}
+
+/**
+ * Given a converter function make an noneable converter function that also coerces
+ * `none` to `undefined`
+ * @param converter - the inner converter function
+ */
+export function noneableAsUndefined<Result, Input = unknown>(
+  converter: ConverterFunction<Result, Input>
+): Converter<Result | undefined, Input | undefined> {
```
Comment (chris-pardy) [inline src/decorators.ts:64]: I think these types are `<Result | undefined, Input | null | undefined>`

File: src/decorators.ts [hunk 10]
```diff
@@ -15,3 +16,56 @@ export function optional<Result, Input = unknown>(
...
 }
+
+/**
+ * Given a converter function make an noneable converter that allows `none`
+ * types to pass through
+ * @param converter - the inner converter function
+ */
+export function noneable<Result, Input = unknown>(
+  converter: ConverterFunction<Result, Input>
+): Converter<Result | None, Input | undefined> {
```
Comment (chris-pardy) [inline src/decorators.ts:27]: I think these types are `<Result | None, Input | None>`

**Comment by pignatenko** [comment]
> @pignatenko approving, but I'd love your take on how this plays into the whole conversation about:
> `t.optional(t.string.default(() => 'empty'))` for instance if I did: `t.noneableAsNull(t.string.default(() => 'empty'))(undefined)` what would I get?

So conceptually speaking default creates a type of 'coercer' really. Functionally speaking default creates a functor in the call chain that replaces missing values with a default. Optional creates a functor in the call chain that allows missing values. 

If we are to think of this functionally as the pure function of a monad Option looks like this: (assume `none` is a cached instance of `None` representing nothingness, and `Some` is just a `.value` holder. 

```
  const option = (value) => {
    if ((value === null) || (value === undefined)) {
      return none;
    }
    return new Some(value)
  }
```
Then default looks like this:

```
  const default = (defaultValue) => (value) => {
    if ((value === null) || (value === undefined)) {
      return new Some(defaultValue);
    }
    return new Some(value)
  }
 ```
 
 Our problem here is `optional` is actually `default(none)`. If we're talking a purely functional interface I would say it's up to the lib user to organize their code to achieve this, it's basically the way it works in all languages as the principle of least surprise.
 
 If we are talking about a declarative OO chained this-return interface, all we need to do is ensure that we create the actual converter pipeline lazily and that `default` is always at the top of the call chain so that we can refuse to re-apply in the special `.optional` case. This is typical of such interfaces.
 
 Right now our problem is that we're mixing our metaphors at the very top level and typical library usage expectations are different in different cases. 
 
e.g.

 I wouldn't expect `t.optional(t.default('empty')(t.string))` to return 'empty' because of just FP.


 I wouldn't expect `t.string.optional.default(() => 'empty')` or `t.string.default(() => 'empty').optional`  to return `undefined` because of OO expectation and how many similar libraries typically have resolved this. (e.g. joi, open API spec, etc).
 
 I support being able to use an FP and an OO interface for declaring things.
